### ISSUE TITLE
tf.function decorator converts ragged tensors to their dense equivalent

### ISSUE THREAD

**Issue by jputila** [tensorflow/tensorflow#26505]
**System information**
- Have I written custom code: No
- OS Platform and Distribution: Windows 10 (will also test on Ubuntu)
- TensorFlow installed from: binary
- TensorFlow version: 2.0 preview
- Python version: 3.6.6
- CUDA/cuDNN version: 10
- GPU model and memory: GTX 1070 8gb

**Describe the current behavior**
When using `@tf.function` a function with a ragged input it will return the dense equivalent of the ragged tensor.  When the decorator is omitted it returns a ragged tensor.

**Describe the expected behavior**
Return a ragged tensor regardless of the use of `@tf.function`

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
@tf.function
def do_nothing(x):
    return x

rt = tf.ragged.constant([[[1, 2, 3], [4]],
                         [[5], [], [6]],
                         [[7]],
                         [[8, 9], [10]]])

do_nothing(rt)
```
```
<tf.Tensor: id=51310, shape=(10,), dtype=int32, numpy=array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])>
```

```
# @tf.function
def do_nothing(x):
    return x

rt = tf.ragged.constant([[[1, 2, 3], [4]],
                         [[5], [], [6]],
                         [[7]],
                         [[8, 9], [10]]])

do_nothing(rt)
```
```
tf.RaggedTensor(values=<tf.RaggedTensor [[1, 2, 3], [4], [5], [], [6], [7], [8, 9], [10]]>, row_splits=tf.Tensor([0 2 5 6 8], shape=(5,), dtype=int64))
```

The ragged tensor documentation hints that this might be intended or necessary, is it?
https://www.tensorflow.org/guide/ragged_tensors#rank_and_ragged_rank

**Comment by alextp** [comment]
@edloper I thought we supported raggedtensors in tf.function due to the changes in nest. Can you take a look?

**Comment by shashvatshahi1998** [comment]
@alextp I want to work on this issue, can you please guide me on this?

**Comment by alextp** [comment]
@shashvatshahi1998 the right way to fix this is to instrument calls to the ragged tensor to tensor function and see when they fire in the tf.function codepath. Whatever's triggering that probably needs to call nest.flatten with expand_composites=True

**Comment by shashvatshahi1998** [comment]
Sorry, @alextp i am not getting what you are trying to say.

**Comment by edloper** [comment]
In third_party/tensorflow/python/eager/function.py, nest is used in several
places to unpack and repack collections of tensors.  Most likely, the
appropriate fix is to add "expand_composites=True" to some (or all?) of
these nest.xyz(...) calls.  This option causes composite tensors to be
expanded into their components.  E.g.:

>>> x = {'a': t1, 'b': RaggedTensor(t2, t3), 'c': tf.SparseTensor(t4, t5,
t6)}
>>> nest.flatten(x)
(t1, RaggedTensor(t2, t3), SparseTensor(t4, t5, t6)
>>> nest.flatten(x, expand_composites=True)
(t1, t2, t3, t4, t5, t6)

Similarly, when expand_composites=True is used with nest.pack_sequence_as,
it will reconstruct the composite tensors from their components.

**Comment by edloper** [comment]
I took an initial stab at this, and got something basic working.  It ends up being a bit more involved than I thought at first (though the basic approach I suggested is the right one).  I'll work on it a bit more and get back to you with an update.

**Comment by shashvatshahi1998** [comment]
@edloper can I also contribute to it.

**Comment by shashvatshahi1998** [comment]
I added expand_composites to all nest.flatten() and nest.pack_sequence_as() what about other functions of nest in the file @edloper

**Comment by edloper** [comment]
Fixed by #40006c3.
